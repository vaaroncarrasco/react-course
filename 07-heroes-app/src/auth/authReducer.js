import { types } from '../types/types';

// * authReducer is just a pure function
export const authReducer = (state = {}, action) => {

    // const state = {
    //     name: 'Fernando',
    //     logged: true
    // }

    switch (action.type) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }

        default:
            return state;
    }

}