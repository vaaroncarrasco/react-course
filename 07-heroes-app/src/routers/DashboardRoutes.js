import React from 'react';
import { Navbar } from '../components/ui/Navbar';
import { Switch, Route, Redirect } from 'react-router-dom';

import { MarvelScreen } from '../components/marvel/MarvelScreen';
import { HeroScreen } from '../components/heroes/HeroScreen';
import { DcScreen } from '../components/dc/DcScreen';
import { SearchScreen } from '../components/search/SearchScreen';

export const DashboardRoutes = () => {

    // ? Only principal routers are wrapped in a <Router></Router> HOC component
    // * Child-router is just a function component. It has only a <Switch></Switch> without <Router></Router>

    return (
        <>
            <Navbar />

            <div className="container mt-2">
                <Switch>
                    <Route exact path="/marvel" component={ MarvelScreen } />
                    <Route exact path="/hero/:heroeId" component={ HeroScreen } />
                    <Route exact path="/dc" component={ DcScreen } />
                    <Route exact path="/search" component={ SearchScreen } />

                    <Redirect to="/marvel" />
                </Switch>
            </div>


        </>
    )
}
