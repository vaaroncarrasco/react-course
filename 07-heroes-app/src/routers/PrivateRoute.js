import React from 'react';
import PropTypes from 'prop-types';

import { Route, Redirect } from 'react-router-dom';


export const PrivateRoute = ({
    isAuthenticated,
    component: Component,
    ...rest
}) => {

    // spread ...rest of props
    // console.log(rest); // rest{ path, location, computedMatch, location }

    // ? Set every time a route passes through PrivateRoute

    localStorage.setItem('lastPath', rest.location.pathname);

    return (
        // { ...rest } passing all props to <Route />
        <Route { ...rest }
            component={ (props) => (
                ( isAuthenticated )
                    ? ( <Component { ...props } /> )
                    : ( <Redirect to="/login" /> )
            )}
        />
    )
}

// Functional Components are functions
PrivateRoute.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}
