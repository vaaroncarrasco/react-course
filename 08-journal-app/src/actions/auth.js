import Swal from "sweetalert2";

import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { noteLogout } from "./notes";
import { finishLoading, startLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ( { user } ) => {

                dispatch(
                    login( user.uid, user.displayName )
                )

                dispatch( finishLoading() );


            }).catch( err => {
                dispatch( finishLoading() );
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }
}


export const startRegisterWithEmailPasswordName = ( email, password, name ) => {

    // ? async fns return (callbacks) => {}
    return ( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword( email, password )
            .then( async ( {user} ) => {

                await user.updateProfile({ displayName: name });

                console.log(user)


                dispatch(
                    login( user.uid, user.displayName )
                )

            }).catch( err => {
                console.log(err);
                Swal.fire('Error', err.message, 'error');
            });

    }

}


export const startGoogleLogin = () => {

    return (dispatch) => {

        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ( { user } ) => {
                dispatch(
                    login( user.uid, user.displayName )
                )
            });

    }

}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {

    // ? Returns an async callback

    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch( logout() );

        dispatch( noteLogout() );
    }
}

export const logout = () => ({
    type: types.logout
})