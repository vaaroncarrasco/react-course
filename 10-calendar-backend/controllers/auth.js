const { response } = require('express');
const bcrypt = require('bcryptjs');
const Usuario = require('../models/Usuario');
const { generarJWT } = require('../helpers/jwt');

const crearUsuario = async( req, res = response ) => {

    const { email, password } = req.body;

    try {

        let usuario = await Usuario.findOne({ email });

        if ( usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'User already exists'
            });
        }

        usuario = new Usuario( req.body );

        // * encrypt password before saving to db
        const salt = bcrypt.genSaltSync();
        usuario.password = bcrypt.hashSync( password, salt );

        await usuario.save();

        // Generate JWT
        const token = await generarJWT( usuario.id, usuario.name );

        res.status(201).json({
            ok: true,
            msg: 'registro',
            uid: usuario.id,
            name: usuario.name,
            token
        });

    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'hable con el admin'
        });
    }
}

const loginUsuario = async ( req, res = response ) => {

    const { email, password } = req.body;

    try {

        const usuario = await Usuario.findOne({ email });

        if ( !usuario ) {
            return res.status(400).json({
                ok: false,
                msg: 'No user found with email'
            });
        }

        // Confirm pass
        const validPassword = bcrypt.compareSync( password, usuario.password );

        if ( !validPassword ) {
            return res.status(400).json({
                ok: false,
                msg: 'Password incorrecto'
            });
        }

        // Generate JWT
        const token = await generarJWT( usuario.id, usuario.name );

        res.json({
            ok: true,
            uid: usuario.id,
            name: usuario.name,
            token
        })



    } catch (error) {
        console.log(error);
        res.status(500).json({
            ok: false,
            msg: 'hable con el admin'
        });
    }

}


const revalidarToken = async( req, res = response ) => {

    const { uid, name } = req;

    // * generate new JWT
    const token = await generarJWT( uid, name );

    res.json({
        ok: true,
        token,
        uid,
        name
    });
}

module.exports = {
    crearUsuario,
    loginUsuario,
    revalidarToken
}