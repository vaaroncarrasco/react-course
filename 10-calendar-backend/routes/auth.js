// * Auth / Users - Route | host + /api/auth

// node uses same lib it has in memory without importing it twice
const { Router } = require('express');
const { check } = require('express-validator');
const { validarCampos } = require('../middlewares/validar-campos');
const { validarJWT } = require('../middlewares/validar-jwt');
const router = Router();


const {
    crearUsuario,
    loginUsuario,
    revalidarToken
} = require('../controllers/auth');


router.post(
    '/new',
    [ // middlewares
        check('name', 'Name is required').not().isEmpty(),
        check('email', 'Email is required').isEmail(),
        check('password', 'Pass must be at least 6 character long').isLength({ min: 6 }),
        validarCampos
    ],
    crearUsuario
);

router.post(
    '/',
    [ // middlewares
        check('email', 'Email is required').isEmail(),
        check('password', 'Pass must be at least 6 character long').isLength({ min: 6 }),
        validarCampos
    ],
    loginUsuario
);

router.get('/renew', validarJWT, revalidarToken);

module.exports = router;