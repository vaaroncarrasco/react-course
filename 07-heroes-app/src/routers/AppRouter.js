import React, { useContext } from 'react'
import {
    BrowserRouter as Router,
    Switch
  } from 'react-router-dom';

import { AuthContext } from '../auth/AuthContext';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';


export const AppRouter = () => {

    // ? Only principal routers are wrapped in a <Router></Router> HOC component with a <Switch></Switch> component inside.
    // * Child-router is just a function component. It has only a <Switch></Switch> without <Router></Router>

    const { user } = useContext(AuthContext);

    return (
        <Router>
            <div>
                {/* We can set global style or html elements in here. They'd appear in every component */}
                <Switch>
                    <PublicRoute
                        exact
                        path="/login"
                        component={ LoginScreen }
                        isAuthenticated={ user.logged }
                    />

                    {/* All routes pass through this non-exact route "/" -> DashboardRoutes */}
                    <PrivateRoute
                        path="/"
                        component={ DashboardRoutes }
                        isAuthenticated={ user.logged }
                    />
                </Switch>
            </div>
        </Router>
    )
}
