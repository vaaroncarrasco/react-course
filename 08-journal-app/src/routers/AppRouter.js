import React, { useEffect, useState } from 'react';
import {
    BrowserRouter as Router,
    Switch,
    Redirect
  } from 'react-router-dom';

import { AuthRouter } from './AuthRouter';
import { JournalScreen } from '../components/journal/JournalScreen';
import { firebase } from '../firebase/firebase-config';
import { useDispatch } from 'react-redux';
import { login } from '../actions/auth';
import { PublicRoute } from './PublicRoute';
import { PrivateRoute } from './PrivateRoute';
import { startLoadingNotes } from '../actions/notes';

export const AppRouter = () => {

    const dispatch = useDispatch();

    // * Check if user is authenticated
    const [checking, setChecking] = useState(true);

    const [isLoggedIn, setIsLoggedIn] = useState(false);

    //? Maintain state using firebase
    useEffect(() => {

        // * onAuthStateChanged -> creates an Observable
        // Observable -> special obj that is triggered depending on events

        // ? Async cause loadNotes() returns promise
        firebase.auth().onAuthStateChanged( async (user) => {

            if ( user?.uid ) {
                dispatch( login(user.uid, user.displayName) );

                setIsLoggedIn( true );

                // * Components should only dispatch and actions do heavy tasks
                dispatch( startLoadingNotes( user.uid ) )

            } else {
                setIsLoggedIn( false );
            }

            setChecking(false);

        })

    }, [dispatch, setChecking, setIsLoggedIn]); // only runs once

    if (checking) {
        return (
            <h1>Wait...</h1>
        )
    }

    return (
        <Router>
            <div>
                <Switch>
                    <PublicRoute
                        isAuthenticated={ isLoggedIn }
                        path="/auth"
                        component={ AuthRouter }
                    />

                    <PrivateRoute
                        isAuthenticated={ isLoggedIn }
                        exact
                        path="/"
                        component={ JournalScreen }
                    />

                    <Redirect to="/auth/login" />

                </Switch>
            </div>
        </Router>
    )
}
