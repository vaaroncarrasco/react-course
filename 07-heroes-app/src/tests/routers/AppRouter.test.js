import React from 'react';
import { mount } from 'enzyme';
import { AppRouter } from '../../routers/AppRouter';
import { AuthContext } from '../../auth/AuthContext';

describe('testing <AppRouter />', () => {

    const contextValue = {
        dispath: jest.fn(),
        user: {
            logged: false
        }
    }

    test('should display login component if not authenticated', () => {

        // ? Testing router with context
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <AppRouter />
            </AuthContext.Provider>
        );

        expect( wrapper ).toMatchSnapshot();

    });

    test('should display heroScreen if authenticated', () => {

        const contextValue = {
            dispath: jest.fn(),
            user: {
                logged: true,
                name: 'EL PEPE'
            }
        }

        // ? Testing router with context
        const wrapper = mount(
            <AuthContext.Provider value={ contextValue }>
                <AppRouter />
            </AuthContext.Provider>
        );

        expect( wrapper.find('.navbar').exists() ).toBe(true);

    });

});