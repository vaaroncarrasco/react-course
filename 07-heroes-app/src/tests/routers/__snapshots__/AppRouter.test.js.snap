// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <AppRouter /> should display login component if not authenticated 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      LoginScreen
    </h1>
    <hr />
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
